### PR TITLE
Implement JWTTokenGenerator

### PR DESCRIPTION
**Pull request by tharindu1st**
## Purpose
Implement JWTTokenGenerator

### CODE REVIEW

File: components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/internal/ServiceReferenceHolder.java [hunk 1]
```diff
@@ -126,4 +132,34 @@ public void setConfig(OAuthConfiguration config) {
...
+        this.privateKey = privateKey;
+    }
+
+    public Certificate getPublicKey() {
+
+        return publicKey;
+    }
+
+    public void setPublicKey(Certificate publicKey) {
+
```
Comment (bhathiya) [inline components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/internal/ServiceReferenceHolder.java]: shall we remove these all extra new lines? or is it automatically added by the code formatter?

File: components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/internal/ConfigurationActivator.java [hunk 2]
```diff
@@ -134,8 +138,16 @@ protected void activate(BundleContext bundleContext) {
                 config.getFileBaseScopes().put(scopeEntry.getKey(), scopeEntry.getValue());
             }
         }
-        config.setFileBaseScopes(populateDefaultFileBaseScopes());
         ServiceReferenceHolder.getInstance().setConfig(config);
+        ServiceReferenceHolder.getInstance().setAuthConfiguration(authConfiguration);
+        try {
+            ServiceReferenceHolder.getInstance().setPrivateKey(org.wso2.carbon.auth.oauth.Utils
```
Comment (bhathiya) [inline components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/internal/ConfigurationActivator.java]: shall we rename `org.wso2.carbon.auth.oauth.Utils` to be `OAuthUtils`?

File: components/auth/org.wso2.carbon.auth.client.registration.rest.api/src/gen/java/org/wso2/carbon/auth/client/registration/rest/api/dto/ApplicationDTO.java [hunk 3]
```diff
@@ -169,6 +172,24 @@ public void setTokenExpireTime(Long tokenExpireTime) {
     this.tokenExpireTime = tokenExpireTime;
   }
 
+  public ApplicationDTO tokenTypeExtension(String tokenTypeExtension) {
+    this.tokenTypeExtension = tokenTypeExtension;
+    return this;
+  }
+
+   /**
```
Comment (bhathiya) [inline components/auth/org.wso2.carbon.auth.client.registration.rest.api/src/gen/java/org/wso2/carbon/auth/client/registration/rest/api/dto/ApplicationDTO.java]: format error

**Review by bhathiya — CHANGES_REQUESTED** [review]

File: features/auth-features/org.wso2.carbon.auth.core.feature/resources/dbscripts/auth/h2.sql [hunk 4]
```diff
@@ -24,6 +24,7 @@ CREATE TABLE IF NOT EXISTS AUTH_OAUTH2_APPLICATION (
             REDIRECT_URI VARCHAR(1024) DEFAULT NULL,
             GRANT_TYPES VARCHAR (1024) DEFAULT NULL,
             APP_STATE VARCHAR (25) DEFAULT 'ACTIVE',
+            TOKEN_TYPE VARCHAR (25) DEFAULT 'Default',
```
Comment (harsha89) [inline features/auth-features/org.wso2.carbon.auth.core.feature/resources/dbscripts/auth/h2.sql:27]: did we change all the sqls?

File: components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        long defaultValidityPeriod = (long) context.getParams().get(OAuthConstants.VALIDITY_PERIOD);
+        long curTimeInMillis = Calendar.getInstance().getTimeInMillis();
+        String sub = (String) context.getParams().get(OAuthConstants.AUTH_USER);
+        String consumerKey = (String) context.getParams().get(OAuthConstants.CLIENT_ID);
+        String issuer = ServiceReferenceHolder.getInstance().getAuthConfigurations().getTokenIssuer();
+        JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder();
+        jwtClaimsSetBuilder.issuer(issuer);
+        jwtClaimsSetBuilder.subject(sub);
+        jwtClaimsSetBuilder.claim(AUTHORIZATION_PARTY, consumerKey);
+        jwtClaimsSetBuilder.issueTime(new Date(curTimeInMillis));
```
Comment (Rajith90) [inline components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java:144]: This value should be in seconds, not in milli seconds

File: components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        jwtClaimsSetBuilder.jwtID(UUID.randomUUID().toString());
+        jwtClaimsSetBuilder.notBeforeTime(new Date(curTimeInMillis));
+        jwtClaimsSetBuilder.claim(SCOPE, scope.toString());
+        if (defaultValidityPeriod < 0) {
+            jwtClaimsSetBuilder.expirationTime(new Date(Long.MAX_VALUE));
+        } else {
+            jwtClaimsSetBuilder.expirationTime(new Date(curTimeInMillis + defaultValidityPeriod * 1000));
```
Comment (Rajith90) [inline components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java:151]: This value should be in seconds, not in milli seconds

File: components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java (hunk 5 above)
Comment (tharindu1st) [reply components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java:144]: The above line converting value to seconds

File: features/auth-features/org.wso2.carbon.auth.core.feature/resources/dbscripts/auth/h2.sql (hunk 4 above)
Comment (tharindu1st) [reply features/auth-features/org.wso2.carbon.auth.core.feature/resources/dbscripts/auth/h2.sql:27]: its not have all the sql scripts yet need to create track from 

https://github.com/wso2/carbon-auth/issues/141

File: components/auth/org.wso2.carbon.auth.client.registration.rest.api/src/gen/java/org/wso2/carbon/auth/client/registration/rest/api/dto/ApplicationDTO.java (hunk 3 above)
Comment (tharindu1st) [reply components/auth/org.wso2.carbon.auth.client.registration.rest.api/src/gen/java/org/wso2/carbon/auth/client/registration/rest/api/dto/ApplicationDTO.java]: auto generated code

**Review by bhathiya — APPROVED** [review]

**Pull request merged**